packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/FallbackJSBundleLoaderTest.kt

```diff
@@ -0,0 +1,169 @@
+
+    assertThat(fallbackLoader.loadScript(null)).isEqualTo("url1")
+
+    verify(delegates[0], times(1)).loadScript(null)
+    verify(delegates[1], never()).loadScript(null)
+
+    assertThat(
+      mLoggingDelegate.logContains(
+        FakeLoggingDelegate.WTF, FallbackJSBundleLoader.TAG, null
+      )
+    ).isFalse
+  }
+
+  @Test
+  fun fallingBackSuccessfully() {
+    val delegates = arrayOf(
+      recoverableLoader("url1", "error1"), successfulLoader("url2"), successfulLoader("url3")
+    )
+
+    val fallbackLoader = FallbackJSBundleLoader(listOf(*delegates))
+
+    assertThat(fallbackLoader.loadScript(null)).isEqualTo("url2")
+
+    verify(delegates[0], times(1)).loadScript(null)
+    verify(delegates[1], times(1)).loadScript(null)
+    verify(delegates[2], never()).loadScript(null)
+
+    assertThat(
+      mLoggingDelegate.logContains(
+        FakeLoggingDelegate.WTF, FallbackJSBundleLoader.TAG, recoverableMsg("error1")
+      )
+    ).isTrue
+  }
+
+  @Test
+  fun fallingbackUnsuccessfully() {
+    val delegates = arrayOf(
+      recoverableLoader("url1", "error1"), recoverableLoader("url2", "error2")
+    )
+
+    val fallbackLoader = FallbackJSBundleLoader(listOf(*delegates))
+
+    try {
+      fallbackLoader.loadScript(null)
+      fail("expect throw")
+    } catch (e: Exception) {
+      assertThat(e).isInstanceOf(RuntimeException::class.java)
+
+      var cause = e.cause
+      val msgs = ArrayList<String?>()
+
+      while (cause != null) {
+        msgs.add(cause.message)
+        cause = cause.cause
+      }
+
+      assertThat(msgs).containsExactly(recoverableMsg("error1"), recoverableMsg("error2"))
+    }
+
+    verify(delegates[0], times(1)).loadScript(null)
+    verify(delegates[1], times(1)).loadScript(null)
+
+    assertThat(
+      mLoggingDelegate.logContains(
+        FakeLoggingDelegate.WTF, FallbackJSBundleLoader.TAG, recoverableMsg("error1")
+      )
+    ).isTrue
+
+    assertThat(
+      mLoggingDelegate.logContains(
+        FakeLoggingDelegate.WTF, FallbackJSBundleLoader.TAG, recoverableMsg("error2")
+      )
+    ).isTrue
+  }
+
+  @Test
+  fun unrecoverable() {
+    val delegates = arrayOf(fatalLoader("url1", "error1"), recoverableLoader("url2", "error2"))
+
+    val fallbackLoader = FallbackJSBundleLoader(listOf(*delegates))
+
+    try {
+      fallbackLoader.loadScript(null)
+      fail("expect throw")
+    } catch (e: Exception) {
+      assertThat(e.message).isEqualTo(fatalMsg("error1"))
+    }
+
+    verify(delegates[0], times(1)).loadScript(null)
+    verify(delegates[1], never()).loadScript(null)
+
+    assertThat(
+      mLoggingDelegate.logContains(
+        FakeLoggingDelegate.WTF, FallbackJSBundleLoader.TAG, null
+      )
+    ).isFalse
+  }
+
+  companion object {
+    private val UNRECOVERABLE: String
+
+    init {
+      val prefix = FallbackJSBundleLoader.RECOVERABLE
+      val first = prefix[0]
+
+      UNRECOVERABLE = prefix.replace(first, (first.code + 1).toChar())
+    }
```

Comment:
This is really not great and we should refactor it a bit.

Having tests rely on the `init{}` block of a `companion object` is really unpredictable.

Let's remove the `companion object` and let's move all the logic to `private` methods and initialize the stuff inside the `setup` method